src/helics/application_api/Subscriptions.cpp

```diff
@@ -81,13 +81,98 @@ void Subscription::handleCallback (Time time)
 
 bool Subscription::isUpdated () const
 {
+    if (hasUpdate)
+    {
+        return true;
+    }
     if (changeDetectionEnabled)
     {
-        return SubscriptionBase::isUpdated ();
+        if (SubscriptionBase::isUpdated ())
+        {
+            if (type == helics_type_t::helicsInvalid)
+            {
+                return true;
+            }
+            auto dv = fed->getValueRaw(id);
+            auto visitor = [&](auto &&arg) -> bool {
+                std::remove_const_t<std::remove_reference_t<decltype (arg)>> newVal;
+                (void)arg;  //suppress VS2015 warning
+                valueExtract (dv, type, newVal);
+                return (changeDetected(lastValue, newVal, delta));
+            };
+            return mpark::visit (visitor, lastValue);
+        }
+        return false;
     }
     else
     {
         return SubscriptionBase::isUpdated ();
     }
 }
+
+bool Subscription::getAndCheckForUpdate ()
+{
+    if (hasUpdate)
+    {
+        return true;
+    }
+    if (changeDetectionEnabled)
+    {
+        if (SubscriptionBase::isUpdated ())
+        {
+            auto dv = fed->getValueRaw (id);
+            if (type == helics_type_t::helicsInvalid)
+            {
+                type = getTypeFromString (fed->getPublicationType (id));
+            }
+            if (type != helics_type_t::helicsInvalid)
+            {
+                auto visitor = [&](auto &&arg) {
+                    std::remove_reference_t<decltype (arg)> newVal;
+                    (void)arg;  //suppress VS2015 warning
+                    valueExtract (dv, type, newVal);
+                    if (changeDetected (lastValue, newVal, delta))
+                    {
+                        lastValue = newVal;
+                        hasUpdate = true;
+                    }
+                };
+                mpark::visit (visitor, lastValue);
+            }
+        }
+    }
+    else
+    {
+        hasUpdate = SubscriptionBase::isUpdated ();
+    }
+
+    return hasUpdate;
+}
+
+size_t Subscription::getRawSize () 
+{ 
+    getAndCheckForUpdate();
+    auto dv=fed->getValueRaw(id);
+    if (dv.empty())
+    {
+        auto out = getValue<std::string>();
+        return out.size();
+    }
+    return dv.size();
+}
+
+size_t Subscription::getStringSize () 
+{ 
+    getAndCheckForUpdate();
+    auto out = getValue<std::string>();
+    return out.size()+1;
+}
+
+size_t Subscription::getVectorSize () 
+{ 
+    getAndCheckForUpdate();
+    auto out = getValue<std::vector<double>>();
```

Comment:
Does this still work if the publication is for a vector of `std::complex<double>` (one of the overloaded options for the `publish` method in the application api)? There might also be some confusion if a federate manually packs a vector of a different data type into the payload.

Reply:
the getValue function will engage all the translation machinery, so any type will get translated to an appropriate vector (doubles, strings, complex, complex vector.  In the case of complex vector it will translate it to a vector of doubles which will have 2x the length of std::vector<double>.    When I started this I thought it would work out a bit cleaner, but my original idea had a lot of corner cases which would fail, so I had to fall back to just doing a full translation to vector<double>

Reply:
I think I might in a later update make the getVectorSize a template function with the default be double.  Then if someone wanted to get the size of the complex<double> vector that would be possible.  But for the moment the primary use case is in the C-api which only has support for vector<double>